x-pack/filebeat/input/s3/s3_integration_test.go

```diff
@@ -0,0 +1,417 @@
+	clientCancel := testSetup(t)
+	defer clientCancel()
+
+	// Simulate input.Context from Filebeat input runner.
+	inputCtx := newInputContext()
+	defer close(inputCtx.Done)
+
+	// Stub outlet for receiving events generated by the input.
+	eventOutlet := newStubOutlet()
+	defer eventOutlet.Close()
+
+	connector := channel.ConnectorFunc(func(_ *common.Config, _ beat.ClientConfig) (channel.Outleter, error) {
+		return eventOutlet, nil
+	})
+
+	in, err := NewInput(cfg, connector, inputCtx)
+	if err != nil {
+		t.Skipf("Skipping: %v", err)
+	}
+	s3Input := in.(*s3Input)
+	defer s3Input.Stop()
+
+	run(s3Input, eventOutlet, t)
+}
+
+func newInputContext() input.Context {
+	return input.Context{
+		Done: make(chan struct{}),
+	}
+}
+
+type stubOutleter struct {
+	sync.Mutex
+	cond   *sync.Cond
+	done   bool
+	Events []beat.Event
+}
+
+func newStubOutlet() *stubOutleter {
+	o := &stubOutleter{}
+	o.cond = sync.NewCond(o)
+	return o
+}
+
+func (o *stubOutleter) waitForEvents(numEvents int) ([]beat.Event, bool) {
+	o.Lock()
+	defer o.Unlock()
+
+	for len(o.Events) < numEvents && !o.done {
+		o.cond.Wait()
+	}
+
+	size := numEvents
+	if size >= len(o.Events) {
+		size = len(o.Events)
+	}
+
+	out := make([]beat.Event, size)
+	copy(out, o.Events)
+	return out, len(out) == numEvents
+}
+
+func (o *stubOutleter) Close() error {
+	o.Lock()
+	defer o.Unlock()
+	o.done = true
+	return nil
+}
+
+func (o *stubOutleter) Done() <-chan struct{} { return nil }
+
+func (o *stubOutleter) OnEvent(event beat.Event) bool {
+	o.Lock()
+	defer o.Unlock()
+	o.Events = append(o.Events, event)
+	o.cond.Broadcast()
+	return !o.done
+}
+
+func TestS3Input(t *testing.T) {
+	cfg := defaultTestConfig()
+
+	runTest(t, cfg, func(input *s3Input, out *stubOutleter, t *testing.T) {
+		config, info := getConfigForTest()
+		if info != "" {
+			t.Skipf("failed to get config for test: %v", info)
+		}
+
+		awsConfig, err := awscommon.GetAWSCredentials(config.AwsConfig)
+		if err != nil {
+
+		}
+		s3BucketRegion := os.Getenv("S3_BUCKET_REGION")
+		if s3BucketRegion == "" {
+			t.Log("S3_BUCKET_REGION is not set, default to us-west-1")
+			s3BucketRegion = "us-west-1"
+		}
+		awsConfig.Region = s3BucketRegion
+		input.awsConfig = awsConfig.Copy()
+		svcSQS := sqs.New(awsConfig)
+
+		// remove old messages from SQS
+		err = input.deleteAllMessages(t, awsConfig, config.QueueURL, int64(config.VisibilityTimeout.Seconds()), svcSQS)
+		if err != nil {
+			t.Fatalf("failed to delete message: %v", err.Error())
+		}
+
+		// upload a sample log file for testing
+		s3BucketNameEnv := os.Getenv("S3_BUCKET_NAME")
+		if s3BucketNameEnv == "" {
+			t.Fatal("failed to get S3_BUCKET_NAME")
+		}
+
+		svcS3 := s3.New(awsConfig)
+		uploadSampleLogFile(t, s3BucketNameEnv, svcS3)
+		time.Sleep(30 * time.Second)
+
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			input.run(svcSQS, svcS3, 300)
+		}()
+
+		time.AfterFunc(10*time.Second, func() { out.Close() })
```

Comment:
Why closing after 10 seconds? Probably the test is not alive for so long :slightly_smiling_face:
Can we defer the close, or close after waiting for events?

Reply:
Yes defer the close works! Thanks!